libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,966 @@
+	m_client{make_unique<JSONTransport>(_in, _out, _logger)},
+	m_handlers{
+		{"cancelRequest", [](auto, auto) {/*don't do anything for now, as we're synchronous*/}},
+		{"$/cancelRequest", [](auto, auto) {/*don't do anything for now, as we're synchronous*/}},
+		{"initialize", bind(&LanguageServer::handle_initialize, this, _1, _2)},
+		{"initialized", {} },
+		{"shutdown", [this](auto, auto) { m_shutdownRequested = true; }},
+		{"workspace/didChangeConfiguration", bind(&LanguageServer::handle_workspace_didChangeConfiguration, this, _1, _2)},
+		{"textDocument/didOpen", bind(&LanguageServer::handle_textDocument_didOpen, this, _1, _2)},
+		{"textDocument/didChange", bind(&LanguageServer::handle_textDocument_didChange, this, _1, _2)},
+		{"textDocument/didClose", [this](auto, Json::Value const& _args) {
+			documentClosed(extractPathFromFileURI(_args["textDocument"]["uri"].asString()).value());
+		}},
+		{"textDocument/definition", bind(&LanguageServer::handle_textDocument_definition, this, _1, _2)},
+		{"textDocument/documentHighlight", bind(&LanguageServer::handle_textDocument_highlight, this, _1, _2)},
+		{"textDocument/references", bind(&LanguageServer::handle_textDocument_references, this, _1, _2)},
+	},
+	m_logger{std::move(_logger)},
+	m_vfs()
+{
+}
+
+void LanguageServer::changeConfiguration(Json::Value const& _settings)
+{
+	if (_settings["evm"].isString())
+		if (auto const evmVersionOpt = EVMVersion::fromString(_settings["evm"].asString()); evmVersionOpt.has_value())
+			m_evmVersion = evmVersionOpt.value();
+
+	if (_settings["remapping"].isArray())
+	{
+		for (auto const& element: _settings["remapping"])
+		{
+			if (element.isString())
+			{
+				if (auto remappingOpt = CompilerStack::parseRemapping(element.asString()); remappingOpt.has_value())
+					m_remappings.emplace_back(move(remappingOpt.value()));
+				else
+					trace("Failed to parse remapping: '"s + element.asString() + "'");
+			}
+		}
+	}
+}
+
+void LanguageServer::documentContentUpdated(string const& _path, std::optional<int> _version, LineColumnRange _range, std::string const& _text)
+{
+	// TODO: all this info is actually unrelated to solidity/lsp specifically except knowing that
+	// the file has updated, so we can  abstract that away and only do the re-validation here.
+	auto file = m_vfs.find(_path);
+	if (!file)
+	{
+		log("LanguageServer: File to be modified not opened \"" + _path + "\"");
+		return;
+	}
+
+	if (_version.has_value())
+		file->setVersion(_version.value());
+
+	file->modify(_range, _text);
+}
+
+void LanguageServer::documentContentUpdated(string const& _path, optional<int> _version, string const& _fullContentChange)
+{
+	auto file = m_vfs.find(_path);
+	if (!file)
+	{
+		log("LanguageServer: File to be modified not opened \"" + _path + "\"");
+		return;
+	}
+
+	if (_version.has_value())
+		file->setVersion(_version.value());
+
+	file->replace(_fullContentChange);
+
+	validate(*file);
+}
+
+void LanguageServer::documentClosed(string const& _path)
```

Comment:
Do we need to support this? Why do we care which documents are open and which are not?

Reply:
I've removed this for now - but at a later point we could use it for dropping internal caches.